Bottom: 7b41ddac234f1809b7912e9d217f568987a032b6
Top:    d2ae6c2fb22fd2dfa3057b7a64aad1606a39a0a0
Author: Andrew Cooper <andrew.cooper@novatechweb.com>
Date:   2018-08-16 09:17:44 -0500

Populate buildbot with gitlab public keys


---
diff --git a/ansible-playbook/buildbot.yml b/ansible-playbook/buildbot.yml
index 509dbc8d..cfb91cca 100644
--- a/ansible-playbook/buildbot.yml
+++ b/ansible-playbook/buildbot.yml
@@ -6,14 +6,20 @@
   become_user: root
   become_method: sudo
   tags:
-  - buildbot
-  roles:
-  - name: Build and start the Buildbot master container
-    role: buildbot-master
-    tags:
+    - buildbot
     - buildbot-master
+  tasks:
+  - name: Scan Gitlab ssh keys
+    command: ssh-keyscan {{gitlab_hostname}}
+    register: gitlab_keys
+  - name: Build and start the Buildbot master container
+    import_role: 
+      name: buildbot-master
+    vars:
+      known_hosts: "{{gitlab_keys.stdout}}"
   - name: Add buildbot user to Gitlab
-    role: gitlab-user
+    import_role: 
+      name: gitlab-user
     vars:
       new_user: buildbot
       new_user_name: Build Robot
@@ -30,7 +36,11 @@
     - buildbot-worker
   vars:
     buildbot_master: "{{ hostvars['buildbot-master'].ip_addr }}"
+    known_hosts: "{{gitlab_keys.stdout}}"
   tasks:
+  - name: Scan Gitlab ssh keys
+    command: ssh-keyscan {{gitlab_hostname}}
+    register: gitlab_keys
   - import_role: 
       name: buildbot-worker-ntel
   - import_role: 
@@ -42,13 +52,13 @@
   become_user: root
   become_method: sudo
   tags:
-  - buildbot
+    - buildbot
   roles:
   - name: Add buildbot_master pubkey to buildbot user
     role: gitlab-sshkey
     vars:
       username: buildbot
-      pubkey: "{{ hostvars.buildbot_master.pubkey }}"
+      pubkey: "{{ hostvars['buildbot-master'].pubkey }}"
       title: "buildbot_master"
   - name: Add buildbot_worker_ntel pubkey to buildbot user
     role: gitlab-sshkey
diff --git a/docker/buildbot-master/buildbot/start.sh b/docker/buildbot-master/buildbot/start.sh
index 85c28248..33b89085 100755
--- a/docker/buildbot-master/buildbot/start.sh
+++ b/docker/buildbot-master/buildbot/start.sh
@@ -8,6 +8,14 @@ BUILDBOT_DATA=${BUILDBOT_DATA-"/buildbot"}
 
 ssh-keygen -q -b ${SSH_BITS} -t ${SSH_TYPE} -N '' -f ${SSH_FILE} 0>&- 1>/dev/null 2>/dev/null
 
+if [ -n "$BUILDBOT_KNOWN_HOSTS_FILE" ] && [ -f "$BUILDBOT_KNOWN_HOSTS_FILE" ]
+then
+    install -v -D \
+        -o `id -u` -g `id -g` -m 0600 \
+        "$BUILDBOT_KNOWN_HOSTS_FILE" \
+        "${HOME}/.ssh/known_hosts"
+fi
+
 ln -v -s -f ${HOME}/buildbot.tac ${HOME}/master.cfg ${HOME}/*.py $BUILDBOT_DATA
 
 # Run original buildbot entrypoint
diff --git a/docker/buildbot-worker-ntel/buildbot/start.sh b/docker/buildbot-worker-ntel/buildbot/start.sh
index f025e661..e7d817e3 100755
--- a/docker/buildbot-worker-ntel/buildbot/start.sh
+++ b/docker/buildbot-worker-ntel/buildbot/start.sh
@@ -10,8 +10,10 @@ ssh-keygen -q -b ${SSH_BITS} -t ${SSH_TYPE} -N '' -f ${SSH_FILE} 0>&- 1>/dev/nul
 
 if [ -n "$BUILDBOT_KNOWN_HOSTS_FILE" ] && [ -f "$BUILDBOT_KNOWN_HOSTS_FILE" ]
 then
-    mkdir -p "${HOME}/.ssh"
-    cp -v "$BUILDBOT_KNOWN_HOSTS_FILE" "${HOME}/.ssh/known_hosts"
+    install -v -D \
+        -o `id -u` -g `id -g` -m 0600 \
+        "$BUILDBOT_KNOWN_HOSTS_FILE" \
+        "${HOME}/.ssh/known_hosts"
 fi
 
 ln -v -s -f ${HOME}/buildbot.tac $BUILDBOT_DATA
diff --git a/docker/buildbot-worker-ptxdist/buildbot/start.sh b/docker/buildbot-worker-ptxdist/buildbot/start.sh
index f025e661..e7d817e3 100755
--- a/docker/buildbot-worker-ptxdist/buildbot/start.sh
+++ b/docker/buildbot-worker-ptxdist/buildbot/start.sh
@@ -10,8 +10,10 @@ ssh-keygen -q -b ${SSH_BITS} -t ${SSH_TYPE} -N '' -f ${SSH_FILE} 0>&- 1>/dev/nul
 
 if [ -n "$BUILDBOT_KNOWN_HOSTS_FILE" ] && [ -f "$BUILDBOT_KNOWN_HOSTS_FILE" ]
 then
-    mkdir -p "${HOME}/.ssh"
-    cp -v "$BUILDBOT_KNOWN_HOSTS_FILE" "${HOME}/.ssh/known_hosts"
+    install -v -D \
+        -o `id -u` -g `id -g` -m 0600 \
+        "$BUILDBOT_KNOWN_HOSTS_FILE" \
+        "${HOME}/.ssh/known_hosts"
 fi
 
 ln -v -s -f ${HOME}/buildbot.tac $BUILDBOT_DATA
diff --git a/roles/buildbot-master/defaults/main.yml b/roles/buildbot-master/defaults/main.yml
index 777499b2..89439ecd 100644
--- a/roles/buildbot-master/defaults/main.yml
+++ b/roles/buildbot-master/defaults/main.yml
@@ -33,6 +33,7 @@ config_hostdir: '{{ docker_restore_config_base_dir }}/{{ container_name }}'
 # environment passed to the container
 container_env:
   BUILDBOT_DATABASE: 'sqlite:///{{ data_path }}/state.sqlite'
+  BUILDBOT_KNOWN_HOSTS_FILE: '{{secrets_path}}/{{secrets_known_hosts}}'
   BUILDBOT_SMTP_HOST: 'relay.dnsexit.com'
   BUILDBOT_SMTP_USER: 'Dean.cox'
   BUILDBOT_SMTP_PASS: 'no5156'
@@ -62,7 +63,7 @@ container_networks:
 # volumes mounted within the container
 container_volumes:
   - '{{ data_volume }}:{{ data_path }}:z'
-  - '{{ config_hostdir }}:{{ secrets_path }}'
+  - '{{ secrets_hostdir }}/{{secrets_known_hosts}}:{{ secrets_path }}/{{secrets_known_hosts}}'
 
 # mount path of the data volume
 data_path: '/buildbot'
@@ -71,7 +72,7 @@ data_path: '/buildbot'
 data_volume: '{{ docker_name_prefix }}buildbot_DV'
 
 # hostname
-hostname: buildbot-worker-ntel
+hostname: '{{ role_name }}'
 
 # build arguments of the image
 image_args:
diff --git a/roles/buildbot-master/tasks/main.yml b/roles/buildbot-master/tasks/main.yml
index b1b843ba..b6d7f58e 100644
--- a/roles/buildbot-master/tasks/main.yml
+++ b/roles/buildbot-master/tasks/main.yml
@@ -10,6 +10,16 @@
   docker_volume:
     name: '{{ data_volume }}'
 
+- name: Create secrets volume
+  file:
+    path: '{{ secrets_hostdir }}'
+    state: directory
+
+- name: Populate known_hosts secret
+  copy:
+    content: "{{ known_hosts }}\n"
+    dest: '{{ secrets_hostdir }}/{{ secrets_known_hosts }}'
+
 - name: Checkout image repo
   git:
     repo: '{{ image_repo }}'
@@ -27,7 +37,7 @@
 - name: Start buildbot container
   docker_container:
     name: '{{ container_name }}'
-    hostname: '{{ buildbot_hostname }}'
+    hostname: '{{ hostname }}'
     image: '{{ image_name }}:{{ docker_image_tag }}'
     networks: '{{ container_networks }}'
     volumes: '{{ container_volumes }}'
@@ -43,6 +53,6 @@
 
 - name: Add buildbot_master host
   add_host:
-    name: buildbot_master
+    name: "{{ hostname }}"
     pubkey: "{{ output.stdout }}"
     ip_addr: "{{ buildbot_ip_addr }}"
diff --git a/roles/buildbot-worker-ntel/tasks/main.yml b/roles/buildbot-worker-ntel/tasks/main.yml
index 32136e2f..f60ef127 100644
--- a/roles/buildbot-worker-ntel/tasks/main.yml
+++ b/roles/buildbot-worker-ntel/tasks/main.yml
@@ -57,7 +57,7 @@
 
 - name: Populate known_hosts secret
   copy:
-    content: '{{ known_hosts }}'
+    content: "{{ known_hosts }}\n"
     dest: '{{ secrets_hostdir }}/{{ secrets_known_hosts }}'
 
 - name: Checkout image repo
diff --git a/roles/buildbot-worker-ptxdist/tasks/main.yml b/roles/buildbot-worker-ptxdist/tasks/main.yml
index b8ed812c..ce43487c 100644
--- a/roles/buildbot-worker-ptxdist/tasks/main.yml
+++ b/roles/buildbot-worker-ptxdist/tasks/main.yml
@@ -55,7 +55,7 @@
 
 - name: Populate known_hosts secret
   copy:
-    content: '{{ known_hosts }}'
+    content: "{{ known_hosts }}\n"
     dest: '{{ secrets_hostdir }}/{{ secrets_known_hosts }}'
 
 - name: Checkout image repo
